src/Adapter/ContainerBuilder.php

```diff
@@ -47,26 +91,183 @@ class ContainerBuilder
+
+        $this->environment = $this->getEnvironment();
+        $this->containerClassName = ucfirst($this->containerName) . 'Container';
+        $this->dumpFile = _PS_CACHE_DIR_ . $this->containerClassName . '.php';
+        $this->containerConfigCache = new ConfigCache($this->dumpFile, $this->isDebug);
+
+        //Necessary to require all annotation classes from Doctrine
+        Setup::createAnnotationMetadataConfiguration([]);
+
+        $container = $this->loadDumpedContainer();
+        if (null === $container) {
+            $container = $this->buildContainer();
         }
+        $this->loadModulesAutoloader();
+
+        return $container;
+    }
 
+    /**
+     * @return ContainerInterface|null
+     */
+    private function loadDumpedContainer()
+    {
+        $container = null;
+        if ($this->containerConfigCache->isFresh()) {
+            require_once $this->dumpFile;
+            $container = new $this->containerClassName();
+        }
+
+        return $container;
+    }
+
+    /**
+     * @return SfContainerBuilder
+     *
+     * @throws \Exception
+     */
+    private function buildContainer()
+    {
         $container = new SfContainerBuilder();
+
+        $this->initParameters($container);
         $container->addCompilerPass(new LegacyCompilerPass());
-        $loader = new YamlFileLoader($container, new FileLocator(__DIR__));
-        $env = $isDebug ? 'dev' : 'prod';
-        $servicesPath = _PS_CONFIG_DIR_ . "services/${name}/services_${env}.yml";
-        $loader->load($servicesPath);
+        $this->initDoctrine($container);
+        $this->loadServices($container);
+
         $container->compile();
 
+        //Dump the container file
         $dumper = new PhpDumper($container);
-        file_put_contents($file, $dumper->dump(array('class' => $containerName)));
+        $this->containerConfigCache->write(
+            $dumper->dump(['class' => $this->containerClassName]),
+            $container->getResources()
+        );
 
         return $container;
     }
+
+    /**
+     * @param SfContainerBuilder $container
+     */
+    private function initParameters(SfContainerBuilder $container)
+    {
+        $parameters = require _PS_ROOT_DIR_ . '/app/config/parameters.php';
+        foreach ($parameters['parameters'] as $parameter => $value) {
+            $container->setParameter($parameter, $value);
+        }
+        $container->setParameter('kernel.bundles', []);
+        $container->setParameter('kernel.root_dir', _PS_ROOT_DIR_ . '/app/');
+        $container->setParameter('kernel.name', 'app');
+        $container->setParameter('kernel.debug', $this->isDebug);
+        $container->setParameter('kernel.environment', $this->environment);
+        $container->setParameter('kernel.cache_dir', _PS_CACHE_DIR_);
+    }
+
+    /**
+     * @param SfContainerBuilder $container
+     *
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    private function initDoctrine(SfContainerBuilder $container)
```

Comment:
imho this should be done in its own compiler pass

Reply:
As I explained in a previous comment this can't be done in a CompilerPass.
The purpose of this method is to add compiler pass to the container, and a CompilerPass which itself adds another CompilerPass is useless because the added one will be purely ignored.

Reply:
in all cases, this should be part of the class ContainerBuilder (SRP).